enhancements/monitoring/multi-tenant-alerting.md

```diff
@@ -0,0 +1,606 @@
+* Ability to use a separate Alertmanager cluster to avoid messing up with the Platform Alertmanager cluster.
+
+#### Non-optimal Alertmanager settings
+
+Users may use non-optimal settings for their alert notifications (such as
+reevaluation of alert groups at high frequency). This may impede the
+performances of Alertmanager since it would consume more resources. It can
+also trigger notification failures if an exteral integration limits the number
+of requests a client IP address can do.
+
+Mitigation
+* Improve the `Alertmanager` CRD to expose new fields enforcing minimum interval values for all associated `AlertmanagerConfig` resources. This would be similar to what exists at the `Prometheus` CRD level for scrape targets with `enforcedSampleLimit` for instance.
+
+#### AlertmanagerConfig resources not being reconciled
+
+The `AlertmanagerConfig` CRD implements schema validation for things that can
+be modeled with the OpenAPI specification. However a
+resource might still not be valid for various reasons:
+* An alerting route contains a sub-route that is invalid (the `route` field has a self-reference to itself which means that it can't be validated at the API level).
+* Credentials (such as API keys) are referenced by secrets, the
+  operator doesn't have permissions to read the secret or the reference
+  is incorrect (wrong name or key).
+
+In such cases, the Prometheus operator discards the invalid
+`AlertmanagerConfig` resource which isn't reconciled in the final Alertmanager
+configuration.
+
+The operator might also be unable to reconcile the AlertmanagerConfig resources temporiraly.
+
+Mitigation
+* The Prometheus operator exposes a validating admission webhook that prevents invalid resources.
+* The Prometheus operator implements the `Status` subresource of the `AlertmanagerConfig` CRD to report whether or not the resource is reconciled or not (see [upstream issue][status-subresource-issue])
+* Users can validate that alerting routing works as expected by generating "fake" alerts triggering the notification system. _Users don't have permissions on the Alertmanager API endpoint so they would have to generate fake alerts from alerting rules themselves. We could also support the ability to craft an alert from the OCP console_.
+
+## Design Details
+
+### Open Questions
+
+1. How can the Dev Console support the UWM Alertmanager?
+
+Users are able to silence alerts from the Dev Console and the console backend
+assumes that the API is served by the
+`alertmanager-main.openshift-monitoring.svc` service. To support the UWM
+Alertmanager configuration, CMO should provide to the console operator the name
+of the Alertmanager service managing the user alerts (either
+`alertmanager-main.openshift-monitoring.svc` or
+`alertmanager.openshift-user-workload-monitoring.svc`). Based on the presence
+of the `openshift_io_alert_source` label, the console backend can decide which
+Alertmanager service should be queried.
```

Comment:
@kyoto @spadgett i'd welcome your inputs on this part. If we allow users to deploy a dedicated Alertmanager for user-defined alerts then this has an impact on the console since the API endpoint to manage the silences depends on the origin of the alert (platform vs. user). Do you see a problem with this?
Implementation-wise, CMO could reuse the `monitoring-shared-config` configmap in the `openshift-managed-config` namespace but I'm open to suggestions.

Reply:
If I understand correctly, this means the Dev Console frontend code will barely change and the Console backend will decide which Alertmanager to proxy the create silence calls to.

When listing the existing silences, both Alertmanagers will need to be queried and the results merged. I guess this merging could be done either by the backend or the frontend, but if the backend is already deciding which Alertmanager to proxy calls to, then maybe it makes sense for the backend to also do the merging.